vdpa/virtio: replace doorbell relay with notifier

kailiangz1 opened the pull request:
Doorbell relay is used before, but it is time consuming. Use notifier instead and use DPDK work to do notify because we can't wait when we process vhost msg, otherwise, deadlock will happend, so, use work thread to finish the job

kailiangz1 committed:
vdpa/virtio: replace doorbell relay with notifier

Doorbell relay is used before, but it is time consuming.
Use notifier instead and use DPDK work to do notify because
we can't wait when we process vhost msg, otherwise, deadlock
will happend, so, use work thread to finish the job

Signed-off-by: Kailiang <kailiangz@nvidia.com>

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
-		DRV_LOG(ERR, "%s virtq doorbell relay failed ret:%d",
-						priv->vdev->device->name, ret);
-		return ret;
+	if (priv->configured){
```
better add a space between ) and {

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
-						priv->vdev->device->name, ret);
-		return ret;
+	if (priv->configured){
+		virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
```
We get this lcore_id using rte_get_next_lcore every time. So do we need to define it as a global var? maybe temp var?

kailiangz1 replied:
we use global to prevent use same core next time

Ch3n60x replied:
ok

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
+	if (priv->configured){
+		virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+		notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+		if(!notify_work) {
```
better add a space after if

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
+		virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+		notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+		if(!notify_work) {
+			DRV_LOG(ERR, "%s vfid %d failed request memory", priv->vdev->device->name, priv->vf_id);
```
Maybe "failed to alloc notify work"?

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -964,6 +829,24 @@ virtio_vdpa_dev_close_work(void *arg)
 	priv->dev_work_flag = VIRTIO_VDPA_DEV_CLOSE_WORK_DONE;
 	return ret;
 }
+static int
```
maybe add a blank line before this

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
+	virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+	notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+	if(!notify_work) {
+		DRV_LOG(ERR, "%s vfid %d failed request memory", vdev->device->name, priv->vf_id);
```
similar comment as above

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
 
+	virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+	notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+	if(!notify_work) {
```
space after if

kailiangz1 replied:
done

Ch3n60x commented on drivers/vdpa/virtio/virtio_vdpa.c:
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
+	DRV_LOG(INFO, "%s vfid %d launch notifier work lcore:%d", vdev->device->name, priv->vf_id, virtio_vdpa_lcore_id);
+	ret = rte_eal_remote_launch(virtio_vdpa_dev_notifier_work, notify_work, virtio_vdpa_lcore_id);
+	if (ret) {
+		DRV_LOG(ERR, "%s vfid %d failed launch notifier work ret:%d lcore:%d", vdev->device->name, priv->vf_id, ret, virtio_vdpa_lcore_id);
```
line 1130 and 1133 seems too long. maybe using two lines?

kailiangz1 replied:
done

Ch3n60x commented:
Did some test. result shows before this commit, set_vring_state could take more than 0.01s for some queue. But after this change, all queues set_vring_state should be below 0.0001s.

kailiangz1 committed:
vdpa/virtio: add new api to modify state interrupt info

Call fd is -1 when pre-config, in order to prevent 2nd restore
modify queue info which takes long time, modify queue's msix
to default value. But dpdk case will different, 2nd restoer time
will longger than kernel

Signed-off-by: Kailiang <kailiangz@nvidia.com>
